components/suggest/src/schema.rs

```diff
@@ -149,15 +158,212 @@ impl ConnectionInitializer for SuggestConnectionInitializer {
         Ok(db.execute_batch(SQL)?)
     }
 
-    fn upgrade_from(&self, _db: &Transaction<'_>, version: u32) -> open_database::Result<()> {
+    fn upgrade_from(&self, tx: &Transaction<'_>, version: u32) -> open_database::Result<()> {
         match version {
             1..=15 => {
                 // Treat databases with these older schema versions as corrupt,
                 // so that they'll be replaced by a fresh, empty database with
                 // the current schema.
                 Err(open_database::Error::Corrupt)
             }
+            16 => {
+                tx.execute(
+                    "
+                    CREATE TABLE dismissed_suggestions (
+                        url_hash INTEGER PRIMARY KEY
+                    ) WITHOUT ROWID;",
+                    (),
+                )?;
+                Ok(())
+            }
             _ => Err(open_database::Error::IncompatibleVersion(version)),
         }
     }
 }
+
+/// Clears the database, removing all suggestions, icons, and metadata.
+pub fn clear_database(db: &Connection) -> rusqlite::Result<()> {
+    db.execute_batch(
+        "
+        DELETE FROM meta;
+        DELETE FROM suggestions;
+        DELETE FROM icons;
+        DELETE FROM yelp_subjects;
+        DELETE FROM yelp_modifiers;
+        DELETE FROM yelp_location_signs;
+        DELETE FROM yelp_custom_details;
+        ",
+    )
+}
+
+fn define_functions(c: &Connection) -> rusqlite::Result<()> {
+    c.create_scalar_function(
+        "hash",
+        1,
+        FunctionFlags::SQLITE_UTF8 | FunctionFlags::SQLITE_DETERMINISTIC,
+        sql_fns::hash,
+    )?;
+    Ok(())
+}
+
+mod sql_fns {
+    use rusqlite::{functions::Context, types::ValueRef, Error, Result};
+    use std::{collections::hash_map::DefaultHasher, hash::Hasher};
+
+    #[inline(never)]
+    pub fn hash(ctx: &Context<'_>) -> Result<Option<i64>> {
+        match ctx.len() {
+            1 => {
+                let raw = ctx.get_raw(0);
+                // This is a deterministic function, which means sqlite
+                // does certain optimizations which means hash() may be called
+                // with a null value even though the query prevents the null
+                // value from actually being used. As a special case, we return
+                // null when the input is NULL. We return NULL instead of zero
+                // because the hash columns are NOT NULL, so attempting to
+                // actually use the null should fail.
+                if raw == ValueRef::Null {
+                    return Ok(None);
+                }
+                match raw.as_str() {
+                    Ok(s) => {
+                        let mut hasher = DefaultHasher::new();
```

Comment:
One potential issue of using the default hasher: if Rust changes the default hasher in the future, all the persisted hash digests in the db would no longer match against to the results using the new hasher. That said, we might want to pick a hasher explicitly here.